In [2]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.metrics import classification_report, confusion_matrix
from sklearn.metrics import accuracy_score

import acquire
import prepare

In [4]:
df_titanic = acquire.get_titanic_data()
df_titanic.head()

passenger_id  survived  pclass     sex   age  sibsp  parch     fare  \
0             0         0       3    male  22.0      1      0   7.2500   
1             1         1       1  female  38.0      1      0  71.2833   
2             2         1       3  female  26.0      0      0   7.9250   
3             3         1       1  female  35.0      1      0  53.1000   
4             4         0       3    male  35.0      0      0   8.0500   

  embarked  class deck  embark_town  alone  
0        S  Third  NaN  Southampton      0  
1        C  First    C    Cherbourg      0  
2        S  Third  NaN  Southampton      1  
3        S  First    C  Southampton      0  
4        S  Third  NaN  Southampton      1

In [5]:
df_titanic = prepare.prep_titanic(df_titanic)
df_titanic.head()
X_train, X_validate, X_test, y_train, y_validate, y_test = prepare.split_data(df_titanic, 'survived')

# print the shapes of the resulting datasets
print(f'Training set shape: {X_train.shape}, {y_train.shape}')
print(f'Validation set shape: {X_validate.shape}, {y_validate.shape}')
print(f'Testing set shape: {X_test.shape}, {y_test.shape}')

Training set shape: (498, 10), (498,)
Validation set shape: (214, 10), (214,)
Testing set shape: (179, 10), (179,)


Fit a K-Nearest Neighbors classifier to your training sample and transform (i.e. make predictions on the training sample)

In [6]:
from sklearn.neighbors import KNeighborsClassifier

# create the classifier
knn = KNeighborsClassifier(n_neighbors=5)

# fit the classifier to the training data
knn.fit(X_train, y_train)

# make predictions on the training data
y_train_pred = knn.predict(X_train)

# compute the accuracy score
accuracy = accuracy_score(y_train, y_train_pred)

# print the accuracy score
print(f'Training accuracy (knn): {accuracy}')


Training accuracy (knn): 0.7469879518072289


/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


Evaluate your results using the model score, confusion matrix, and classification report.

In [7]:
from sklearn.metrics import confusion_matrix, classification_report

# compute the accuracy score on the test data
accuracy = knn.score(X_test, y_test)

# make predictions on the test data
y_pred = knn.predict(X_test)

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# compute the classification report
cr = classification_report(y_test, y_pred)

# print the results
print(f'Test accuracy: {accuracy}')
print(f'Confusion matrix:\n{cm}')
print(f'Classification report:\n{cr}')


Test accuracy: 0.6312849162011173
Confusion matrix:
[[83 27]
 [39 30]]
Classification report:
              precision    recall  f1-score   support

           0       0.68      0.75      0.72       110
           1       0.53      0.43      0.48        69

    accuracy                           0.63       179
   macro avg       0.60      0.59      0.60       179
weighted avg       0.62      0.63      0.62       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. 

Print and clearly label the following: Accuracy, true positive rate, false positive rate, true negative rate, false negative rate, precision, recall, f1-score, and support.

In [8]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# make predictions on the test data
y_pred = knn.predict(X_test)

# compute and print the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print(f'Accuracy: {accuracy:.2f}')

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# extract the true positive, false positive, true negative, and false negative rates from the confusion matrix
tn, fp, fn, tp = cm.ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)

# print the true positive rate, false positive rate, true negative rate, and false negative rate
print(f'True Positive Rate (Recall): {tpr:.2f}')
print(f'False Positive Rate: {fpr:.2f}')
print(f'True Negative Rate: {tnr:.2f}')
print(f'False Negative Rate: {fnr:.2f}')

# compute and print the classification report
cr = classification_report(y_test, y_pred)
print(f'Classification Report:\n{cr}')


Accuracy: 0.63
True Positive Rate (Recall): 0.43
False Positive Rate: 0.25
True Negative Rate: 0.75
False Negative Rate: 0.57
Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.75      0.72       110
           1       0.53      0.43      0.48        69

    accuracy                           0.63       179
   macro avg       0.60      0.59      0.60       179
weighted avg       0.62      0.63      0.62       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


Run through steps 1-3 setting k to 10


In [9]:
from sklearn.neighbors import KNeighborsClassifier

# create the classifier
knn = KNeighborsClassifier(n_neighbors=10)

# fit the classifier to the training data
knn.fit(X_train, y_train)

# make predictions on the training data
y_train_pred = knn.predict(X_train)

# compute the accuracy score
accuracy = accuracy_score(y_train, y_train_pred)

# print the accuracy score
print(f'Training accuracy (knn): {accuracy}')

Training accuracy (knn): 0.6947791164658634


/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [19]:
# compute the accuracy score on the test data
accuracy = knn.score(X_test, y_test)

# make predictions on the test data
y_pred1 = knn.predict(X_test)

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# compute the classification report
cr = classification_report(y_test, y_pred)

# print the results
print(f'Test accuracy: {accuracy}')
print(f'Confusion matrix:\n{cm}')
print(f'Classification report:\n{cr}')

Test accuracy: 0.6927374301675978
Confusion matrix:
[[101   9]
 [ 46  23]]
Classification report:
              precision    recall  f1-score   support

           0       0.69      0.92      0.79       110
           1       0.72      0.33      0.46        69

    accuracy                           0.69       179
   macro avg       0.70      0.63      0.62       179
weighted avg       0.70      0.69      0.66       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. 

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# make predictions on the test data
y_pred1 = knn.predict(X_test)

# compute and print the accuracy score
accuracy = accuracy_score(y_test, y_pred1)
print(f'Accuracy: {accuracy:.2f}')

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred1)

# extract the true positive, false positive, true negative, and false negative rates from the confusion matrix
tn, fp, fn, tp = cm.ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)

# print the true positive rate, false positive rate, true negative rate, and false negative rate
print(f'True Positive Rate (Recall): {tpr:.2f}')
print(f'False Positive Rate: {fpr:.2f}')
print(f'True Negative Rate: {tnr:.2f}')
print(f'False Negative Rate: {fnr:.2f}')

# compute and print the classification report
cr = classification_report(y_test, y_pred)
print(f'Classification Report:\n{cr}')

Accuracy: 0.69
True Positive Rate (Recall): 0.33
False Positive Rate: 0.08
True Negative Rate: 0.92
False Negative Rate: 0.67
Classification Report:
              precision    recall  f1-score   support

           0       0.69      0.92      0.79       110
           1       0.72      0.33      0.46        69

    accuracy                           0.69       179
   macro avg       0.70      0.63      0.62       179
weighted avg       0.70      0.69      0.66       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


Run through steps 1-3 setting k to 20

In [15]:
from sklearn.neighbors import KNeighborsClassifier

# create the classifier
knn = KNeighborsClassifier(n_neighbors=20)

# fit the classifier to the training data
knn.fit(X_train, y_train)

# make predictions on the training data
y_train_pred = knn.predict(X_train)

# compute the accuracy score
accuracy = accuracy_score(y_train, y_train_pred)

# print the accuracy score
print(f'Training accuracy (knn): {accuracy}')

Training accuracy (knn): 0.6746987951807228


/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


In [16]:
# compute the accuracy score on the test data
accuracy = knn.score(X_test, y_test)

# make predictions on the test data
y_pred2 = knn.predict(X_test)

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred2)

# compute the classification report
cr = classification_report(y_test, y_pred2)

# print the results
print(f'Test accuracy: {accuracy}')
print(f'Confusion matrix:\n{cm}')
print(f'Classification report:\n{cr}')

Test accuracy: 0.6927374301675978
Confusion matrix:
[[101   9]
 [ 46  23]]
Classification report:
              precision    recall  f1-score   support

           0       0.69      0.92      0.79       110
           1       0.72      0.33      0.46        69

    accuracy                           0.69       179
   macro avg       0.70      0.63      0.62       179
weighted avg       0.70      0.69      0.66       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. 

In [17]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# make predictions on the test data
y_pred2 = knn.predict(X_test)

# compute and print the accuracy score
accuracy = accuracy_score(y_test, y_pred2)
print(f'Accuracy: {accuracy:.2f}')

# compute the confusion matrix
cm = confusion_matrix(y_test, y_pred2)

# extract the true positive, false positive, true negative, and false negative rates from the confusion matrix
tn, fp, fn, tp = cm.ravel()
tpr = tp / (tp + fn)
fpr = fp / (fp + tn)
tnr = tn / (tn + fp)
fnr = fn / (fn + tp)

# print the true positive rate, false positive rate, true negative rate, and false negative rate
print(f'True Positive Rate (Recall): {tpr:.2f}')
print(f'False Positive Rate: {fpr:.2f}')
print(f'True Negative Rate: {tnr:.2f}')
print(f'False Negative Rate: {fnr:.2f}')

# compute and print the classification report
cr = classification_report(y_test, y_pred)
print(f'Classification Report:\n{cr}')

Accuracy: 0.69
True Positive Rate (Recall): 0.33
False Positive Rate: 0.08
True Negative Rate: 0.92
False Negative Rate: 0.67
Classification Report:
              precision    recall  f1-score   support

           0       0.69      0.92      0.79       110
           1       0.72      0.33      0.46        69

    accuracy                           0.69       179
   macro avg       0.70      0.63      0.62       179
weighted avg       0.70      0.69      0.66       179



/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


### What are the differences in the evaluation metrics? Which performs better on your in-sample data? Why?

The two models have different evaluation metrics. The first model with k=10 has a higher training accuracy of 0.6948 and a lower test accuracy of 0.6648. The second model with k=20 has a lower training accuracy of 0.6747 and a higher test accuracy of 0.6927. 

In terms of the classification report, the second model with k=20 has a higher precision, recall, and f1-score for the positive class (1) indicating that it is better at predicting the survival of passengers who did not survive in the Titanic disaster. However, the first model with k=10 has a higher precision, recall, and f1-score for the negative class (0) indicating that it is better at predicting the survival of passengers who survived.

Overall, the second model with k=20 performs slightly better on the in-sample data as it has a higher test accuracy and better performance on the positive class. However, it's important to keep in mind that the performance of the models can vary depending on the specific data being used and other factors such as the choice of distance metric.

I would use the k=5 because it has the highest accuracy score.


Which model performs best on our out-of-sample data from validate?

In [22]:

# Train and evaluate the first model with k=10
knn1 = KNeighborsClassifier(n_neighbors=10)
knn1.fit(X_train, y_train)
y_pred1 = knn1.predict(X_validate)

print("k set at 10\n")
print("Validation accuracy:", accuracy_score(y_validate, y_pred1))
print("Confusion matrix:\n", confusion_matrix(y_validate, y_pred1))
print("Classification report:\n", classification_report(y_validate, y_pred1))

# Train and evaluate the second model with k=20
knn2 = KNeighborsClassifier(n_neighbors=20)
knn2.fit(X_train, y_train)
y_pred2 = knn2.predict(X_validate)

print("\nk set at 20\n")
print("Validation accuracy:", accuracy_score(y_validate, y_pred2))
print("Confusion matrix:\n", confusion_matrix(y_validate, y_pred2))
print("Classification report:\n", classification_report(y_validate, y_pred2))

# Determine which model performs best on out-of-sample data
if accuracy_score(y_validate, y_pred1) > accuracy_score(y_validate, y_pred2):
    print("Model with k=10 performs best on out-of-sample data")
else:
    print("Model with k=20 performs best on out-of-sample data")



k set at 10

Validation accuracy: 0.616822429906542
Confusion matrix:
 [[114  18]
 [ 64  18]]
Classification report:
               precision    recall  f1-score   support

           0       0.64      0.86      0.74       132
           1       0.50      0.22      0.31        82

    accuracy                           0.62       214
   macro avg       0.57      0.54      0.52       214
weighted avg       0.59      0.62      0.57       214


k set at 20

Validation accuracy: 0.6588785046728972
Confusion matrix:
 [[123   9]
 [ 64  18]]
Classification report:
               precision    recall  f1-score   support

           0       0.66      0.93      0.77       132
           1       0.67      0.22      0.33        82

    accuracy                           0.66       214
   macro avg       0.66      0.58      0.55       214
weighted avg       0.66      0.66      0.60       214

Model with k=20 performs best on out-of-sample data


/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
/usr/local/anaconda3/lib/python3.9/site-packages/sklearn/neighbors/_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. 